File: security-platform/agents/dotnet/SecurityPlatform.Agent/Redaction.cs
using System.Text.RegularExpressions;

namespace SecurityPlatform.Agent;

/// <summary>
/// PII Redaction utility
/// </summary>
public class Redaction
{
    private static readonly List<Regex> DefaultPatterns = new()
    {
        new Regex(@"(?i)(password|passwd|pwd)\s*[:=]\s*([^\s,}]+)", RegexOptions.IgnoreCase),
        new Regex(@"(?i)(api[_-]?key|apikey)\s*[:=]\s*([^\s,}]+)", RegexOptions.IgnoreCase),
        new Regex(@"(?i)(token|bearer)\s*[:=]\s*([^\s,}]+)", RegexOptions.IgnoreCase),
        new Regex(@"\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b"), // Credit card
        new Regex(@"\b\d{3}-\d{2}-\d{4}\b"), // SSN
    };

    private readonly List<Regex> _patterns;

    public Redaction(List<string>? customRules = null)
    {
        _patterns = new List<Regex>(DefaultPatterns);
        if (customRules != null)
        {
            foreach (var rule in customRules)
            {
                try
                {
                    _patterns.Add(new Regex(rule, RegexOptions.IgnoreCase));
                }
                catch
                {
                    // Invalid regex, skip
                }
            }
        }
    }

    public string Redact(string? input)
    {
        if (string.IsNullOrEmpty(input))
        {
            return input ?? string.Empty;
        }

        var result = input;
        foreach (var pattern in _patterns)
        {
            result = pattern.Replace(result, "[REDACTED]");
        }
        return result;
    }

    public Dictionary<string, object> Redact(Dictionary<string, object>? data)
    {
        if (data == null)
        {
            return new Dictionary<string, object>();
        }

        var redacted = new Dictionary<string, object>();
        foreach (var entry in data)
        {
            if (ShouldRedactKey(entry.Key))
            {
                redacted[entry.Key] = "[REDACTED]";
            }
            else if (entry.Value is string strValue)
            {
                redacted[entry.Key] = Redact(strValue);
            }
            else if (entry.Value is Dictionary<string, object> dictValue)
            {
                redacted[entry.Key] = Redact(dictValue);
            }
            else
            {
                redacted[entry.Key] = entry.Value;
            }
        }
        return redacted;
    }

    private static bool ShouldRedactKey(string key)
    {
        var lowerKey = key.ToLowerInvariant();
        return lowerKey.Contains("password", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("passwd", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("pwd", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("api_key", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("apikey", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("token", StringComparison.OrdinalIgnoreCase) ||
               lowerKey.Contains("secret", StringComparison.OrdinalIgnoreCase);
    }
}

File: security-platform/agents/ios/Sources/SecurityPlatform/Redaction.swift
import Foundation

/// PII Redaction utility
public class Redaction {
    private let patterns: [NSRegularExpression]
    
    private static let defaultPatterns: [String] = [
        "(?i)(password|passwd|pwd)\\s*[:=]\\s*([^\\s,}]+)",
        "(?i)(api[_-]?key|apikey)\\s*[:=]\\s*([^\\s,}]+)",
        "(?i)(token|bearer)\\s*[:=]\\s*([^\\s,}]+)",
        "\\b\\d{4}[\\s-]?\\d{4}[\\s-]?\\d{4}[\\s-]?\\d{4}\\b", // Credit card
        "\\b\\d{3}-\\d{2}-\\d{4}\\b" // SSN
    ]
    
    public init(customRules: [String]? = nil) {
        var compiledPatterns: [NSRegularExpression] = []
        
        for pattern in Self.defaultPatterns {
            if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
                compiledPatterns.append(regex)
            }
        }
        
        if let customRules = customRules {
            for rule in customRules {
                if let regex = try? NSRegularExpression(pattern: rule, options: .caseInsensitive) {
                    compiledPatterns.append(regex)
                }
            }
        }
        
        self.patterns = compiledPatterns
    }
    
    public func redact(_ input: String) -> String {
        guard !input.isEmpty else { return input }
        
        var result = input
        for pattern in patterns {
            result = pattern.stringByReplacingMatches(
                in: result,
                options: [],
                range: NSRange(location: 0, length: result.utf16.count),
                withTemplate: "[REDACTED]"
            )
        }
        return result
    }
    
    public func redact(_ data: [String: Any]) -> [String: Any] {
        var redacted: [String: Any] = [:]
        
        for (key, value) in data {
            if shouldRedactKey(key) {
                redacted[key] = "[REDACTED]"
            } else if let stringValue = value as? String {
                redacted[key] = redact(stringValue)
            } else if let dictValue = value as? [String: Any] {
                redacted[key] = redact(dictValue)
            } else {
                redacted[key] = value
            }
        }
        
        return redacted
    }
    
    private func shouldRedactKey(_ key: String) -> Bool {
        let lowerKey = key.lowercased()
        return lowerKey.contains("password") ||
               lowerKey.contains("passwd") ||
               lowerKey.contains("pwd") ||
               lowerKey.contains("api_key") ||
               lowerKey.contains("apikey") ||
               lowerKey.contains("token") ||
               lowerKey.contains("secret")
    }
}
